### PR TITLE
Ticket6841 queue actions individually

### PR DESCRIPTION
**Pull request by JamesKingWork**
### Description of work

Add the ability to play, pause and stop the execution of actions dynamically in the script generator.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/6841

### Acceptance criteria

*List the acceptance criteria for the PR. The aim is provide information to help the reviewer*

### Unit tests

See tests in uk.ac.stfc.isis.ibex.scriptgenerator.tests

### System tests

*Mention any automated tests or manual tests that you have added or modified, if applicable. The aim is provide information to help the reviewer*

### Documentation

https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IBEX-Script-Generator-Dynamic-Scripting

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Is there associated PR for the [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)?

### Final Steps
- [ ] Reviewer has also merged the [release notes](https://github.com/ISISComputingGroup/IBEX/blob/master/release_notes/ReleaseNotes_Upcoming.md)

### CODE REVIEW

**Review by ThomasLohnert — CHANGES_REQUESTED** [review]
Functional review:
All interactions feel very robust and intuitive in design (to me at least). Great work!!

Some things I found:

- As discussed- very short actions break execution, e.g. an action that only prints something (no sleeps etc.). The script generator will think it is executing this action forever. The status update that the script server it has finished possibly comes before the script generator starts listening
- Noticed a bug with the active button highlighting becoming inaccurate. To reproduce:
    1. Have an action that contains a sleep
    2. Pause and Resume execution while it is sleeping
    3. The "resume" instruction reaches the script server before it is in paused state and is ignored, but the buttons in the script generator view think it's now "running"
    4. From here, you also cannot resume execution by pressing "Run" again, only after you press "Pause" again first.
- Pause logos in table are basically unrecognizable, could we scale up the font size for them? Ideally we'd use the same images for the buttons but I appreciate doing this inside a table may be a pain. (This is a relatively minor point, as there are many other indicators for server status, so don't spend tons of time on this. This is more about the level of polish)

**Comment by JamesKingWork** [comment]
> As discussed- very short actions break execution, e.g. an action that only prints something (no sleeps etc.). The script generator will think it is executing this action forever. The status update that the script server it has finished possibly comes before the script generator starts listening

It turns out the script status property changed isn't fired by the Nicos model when scripts change very fast and we are listening to it in time. The scriptName and currentlyExecutingScript properties do get changed and fired, so I have added listening for these properties and slightly refactored the checking for if the script has finished in https://github.com/ISISComputingGroup/ibex_gui/pull/1375/commits/3a4a6e039b0a8b839bc7583a91c00a38bea54c54.

**Comment by JamesKingWork** [comment]
> Noticed a bug with the active button highlighting becoming inaccurate. To reproduce:
>
>    Have an action that contains a sleep
>    Pause and Resume execution while it is sleeping
>    The "resume" instruction reaches the script server before it is in paused state and is ignored, but the buttons in the script generator view think it's now "running"
>    From here, you also cannot resume execution by pressing "Run" again, only after you press "Pause" again first.

Originally, pressing resume in the paused state presumed that the pause had completed successfully. With https://github.com/ISISComputingGroup/ibex_gui/pull/1375/commits/b2221aac22a190469937e51bbffba055d4d70eb3 we now wait for Nicos to notify us a script in "in break" before being able to resume.

**Comment by JamesKingWork** [comment]
> Pause logos in table are basically unrecognizable, could we scale up the font size for them? Ideally we'd use the same images for the buttons but I appreciate doing this inside a table may be a pain. (This is a relatively minor point, as there are many other indicators for server status, so don't spend tons of time on this. This is more about the level of polish)

As of https://github.com/ISISComputingGroup/ibex_gui/pull/1375/commits/95236cbc9c14e9da8e87be64bb83c79c8d2a0b41 the table now uses the images also used by the buttons. I have also made it so the resum button switches the image when paused. This commit also contains fixes for the previous commits tests which were missed in the previous commit.

File: base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [DEFAULT]
-uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.scriptGenerator,uk.ac.stfc.isis.ibex.client.e4.product.perspective.reflectometry
+uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.reflectometry
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini]: No, my bad, I've now added it back into the ini

**Review by ThomasLohnert — APPROVED** [review]

**Pull request merged**